src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -1050,16 +1050,37 @@ public static bool TryDecode(ImmutableArray<byte> bytes)
 
         /// <summary>
         /// Given an input string changes it to be acceptable as a part of a type name.
-        /// For now we will simply replace '.' with '_'as the most common case.
         /// </summary>
         internal static string MangleForTypeNameIfNeeded(string moduleName)
-        {
-            // TODO: it may make sense to strenthen this algorithm 
-            //       to result in 1-1 mapping to reduce chances of
-            //       producing matching results for distinct original strings
-            var result = moduleName.Replace('.', '_');
+        {            
+            var s = new StringBuilder();
+            foreach (var c in moduleName)
+            {
+                //Any character that can appear in an identifier maps to itself, except Q and _
+                var qOr_ = (c == 'Q' || c == '_');
+                if ((!qOr_) && char.IsLetterOrDigit(c))
+                {
+                    s.Append(c);
+                    continue;
+                }
+                switch (c)
+                {
+                    case ('Q'): //Q maps to QQ
+                        s.Append("QQ");
+                        break;
+                    case ('_'): //_ maps to Q_
+                        s.Append("Q_");
+                        break;
+                    case ('.'): //. maps to _
+                        s.Append('_');
+                        break;
+                    default: //any other character maps to Q followed by four hex digits for its unicode value.
+                        s.AppendFormat("{0}{1:X}", 'Q', c);
```

Comment:
According to ECMA the only character that's not allowed in a type name is '\0'. So technically '.' is valid as well, but since the legacy metadata reader API (IMetadataImport) don't work well with them we want to avoid it to not confuse tools that use these APIs.